src/decoder/lattice-faster-decoder-combine-bucketqueue.cc

```diff
@@ -0,0 +1,1133 @@
+  cost_offsets_.resize(frame + 1, 0.0);
+  cost_offsets_[frame] = cost_offset;
+
+  // Iterator the "cur_queue_" to process non-emittion and emittion arcs in fst.
+  Token *tok = NULL;
+  int32 num_toks_processed = 0;
+  int32 max_active = config_.max_active;
+  for (; num_toks_processed < max_active && (tok = cur_queue_.Pop()) != NULL;
+       num_toks_processed++) {
+    BaseFloat cur_cost = tok->tot_cost;
+    StateId state = tok->state_id;
+    if (cur_cost > cur_cutoff) { // Don't bother processing successors.
+      break;  // This is a priority queue. The following tokens will be worse
+    } else if (cur_cost + adaptive_beam < cur_cutoff) {
+      cur_cutoff = cur_cost + adaptive_beam; // a tighter boundary
+    }
+    // If "tok" has any existing forward links, delete them,
+    // because we're about to regenerate them.  This is a kind
+    // of non-optimality (remember, this is the simple decoder),
+    DeleteForwardLinks(tok);  // necessary when re-visiting
+    tok->links = NULL;
+    for (fst::ArcIterator<FST> aiter(*fst_, state);
+         !aiter.Done();
+         aiter.Next()) {
+      const Arc &arc = aiter.Value();
+      bool changed;
+      if (arc.ilabel == 0) {  // propagate nonemitting
+        BaseFloat graph_cost = arc.weight.Value();
+        BaseFloat tot_cost = cur_cost + graph_cost;
+        if (tot_cost < cur_cutoff) {
+          Token *new_tok = FindOrAddToken(arc.nextstate, frame, tot_cost,
+                                          tok, &prev_toks_, &changed);
+
+          // Add ForwardLink from tok to new_tok. Put it on the head of
+          // tok->link list
+          tok->links = new ForwardLinkT(new_tok, 0, arc.olabel,
+                                        graph_cost, 0, tok->links);
+          
+          // "changed" tells us whether the new token has a different
+          // cost from before, or is new.
+          if (changed && !new_tok->in_queue) {
+            cur_queue_.Push(new_tok);
+          }
+        }
+      } else {  // propagate emitting
+        BaseFloat graph_cost = arc.weight.Value(),
+                  ac_cost = cost_offset - decodable->LogLikelihood(frame, arc.ilabel),
+                  cur_cost = tok->tot_cost,
+                  tot_cost = cur_cost + ac_cost + graph_cost;
+        if (tot_cost > next_cutoff) continue;
+        else if (tot_cost + adaptive_beam < next_cutoff) {
+          next_cutoff = tot_cost + adaptive_beam;  // a tighter boundary for emitting
+        }
+
+        // no change flag is needed
+        Token *next_tok = FindOrAddToken(arc.nextstate, frame + 1, tot_cost,
+                                         tok, &cur_toks_, NULL);
+        // Add ForwardLink from tok to next_tok. Put it on the head of tok->link
+        // list
+        tok->links = new ForwardLinkT(next_tok, arc.ilabel, arc.olabel,
+                                      graph_cost, ac_cost, tok->links);
+        if (best_token_in_next_frame_ == NULL ||
```

Comment:
This code can be optimized by combining with the code around line 873.   Move the
```if (tot_cost + adaptive_beam < next_cutoff) {
          next_cutoff = tot_cost + adaptive_beam;  // a tighter boundary for emitting
        }
```
to here,  and inside the if-statement, do:
```
    best_token_in_next_frame_ = next_tok
```
Then you don't need this extra if-statement.